libyul/backends/evm/EVMDialect.cpp

```diff
@@ -342,6 +343,21 @@ std::map<YulName, BuiltinFunctionForEVM> createBuiltins(langutil::EVMVersion _ev
 				_assembly.appendImmutable(formatLiteral(std::get<Literal>(_call.arguments.front())));
 			}
 		));
+		builtins.emplace(createFunction(
+			"dataloadn",
+			1,
+			1,
+			SideEffects{},
+			{LiteralKind::String},
+			[](
+				FunctionCall const& _call,
+				AbstractAssembly& _assembly,
+				BuiltinContext&
+			) {
+				yulAssert(_call.arguments.size() == 1, "");
+				_assembly.appendDataLoadN(std::stoul(formatLiteral(std::get<Literal>(_call.arguments.front()))));
```

Comment:
You should validate that the literal is in the right range. [We have a place for such checks in `AsmAnalysis`](https://github.com/ethereum/solidity/blob/v0.8.27/libyul/AsmAnalysis.cpp#L406-L427). Could also be an unimplemented assert here with a TODO to do it later, but just adding the check there will be basically the same effort.

Also, it should really be a number, not a string, syntactically. As a bonus, if you declare it as `LiteralKind::Number` rather than `LiteralKind::String`, you won't have to convert it yourself and worry about handling special cases other than it just being out of range:
```suggestion
				_assembly.appendDataLoadN(std::get<Literal>(_call.arguments.front()).value.value());
```

IIRC you had some problems with the literal here back when you started your EOF implementation, but I think it should work as a number just fine now. `memoryguard()` does accept an integer already and also @clonker has rationalized this whole business with literals since then (#15112) so it should no longer be an issue.